plugins/logs/plugin.go

```diff
@@ -51,13 +52,144 @@ type EventV1 struct {
 	RequestedBy string                  `json:"requested_by"`
 	Timestamp   time.Time               `json:"timestamp"`
 	Metrics     map[string]interface{}  `json:"metrics,omitempty"`
+
+	inputAST ast.Value
 }
 
 // BundleInfoV1 describes a bundle associated with a decision log event.
 type BundleInfoV1 struct {
 	Revision string `json:"revision,omitempty"`
 }
 
+// AST returns the BundleInfoV1 as an AST value
+func (b *BundleInfoV1) AST() ast.Value {
+	result := ast.NewObject()
+	if len(b.Revision) > 0 {
+		result.Insert(ast.StringTerm("revision"), ast.StringTerm(b.Revision))
+	}
+	return result
+}
+
+// Key ast.Term values for the Rego AST representation of the EventV1
+var labelsKey = ast.StringTerm("labels")
+var decisionIDKey = ast.StringTerm("decision_id")
+var revisionKey = ast.StringTerm("revision")
+var bundlesKey = ast.StringTerm("bundles")
+var pathKey = ast.StringTerm("path")
+var queryKey = ast.StringTerm("query")
+var inputKey = ast.StringTerm("input")
+var resultKey = ast.StringTerm("result")
+var erasedKey = ast.StringTerm("erased")
+var errorKey = ast.StringTerm("error")
+var requestedByKey = ast.StringTerm("requested_by")
+var timestampKey = ast.StringTerm("timestamp")
+var metricsKey = ast.StringTerm("metrics")
+
+// AST returns the Rego AST representation for a given EventV1 object.
+// This avoids having to round trip through JSON while applying a decision log
+// mask policy to the event.
+func (e *EventV1) AST() (ast.Value, error) {
+	var err error
+	event := ast.NewObject()
+
+	if e.Labels != nil {
+		labelsObj := ast.NewObject()
+		for k, v := range e.Labels {
+			labelsObj.Insert(ast.StringTerm(k), ast.StringTerm(v))
+		}
+		event.Insert(labelsKey, ast.NewTerm(labelsObj))
+	} else {
+		event.Insert(labelsKey, ast.NullTerm())
+	}
+
+	event.Insert(decisionIDKey, ast.StringTerm(e.DecisionID))
+
+	if len(e.Revision) > 0 {
+		event.Insert(revisionKey, ast.StringTerm(e.Revision))
+	}
+
+	if len(e.Bundles) > 0 {
+		bundlesObj := ast.NewObject()
+		for k, v := range e.Bundles {
+			bundlesObj.Insert(ast.StringTerm(k), ast.NewTerm(v.AST()))
+		}
+		event.Insert(bundlesKey, ast.NewTerm(bundlesObj))
+	}
+
+	if len(e.Path) > 0 {
+		event.Insert(pathKey, ast.StringTerm(e.Path))
+	}
+
+	if len(e.Query) > 0 {
+		event.Insert(queryKey, ast.StringTerm(e.Query))
+	}
+
+	if e.Input != nil {
+		if e.inputAST == nil {
+			e.inputAST, err = roundtripJSONToAST(e.Input)
+			if err != nil {
+				return nil, err
+			}
+		}
+		event.Insert(inputKey, ast.NewTerm(e.inputAST))
+	}
+
+	if e.Result != nil {
+		results, err := roundtripJSONToAST(e.Result)
+		if err != nil {
+			return nil, err
+		}
+		event.Insert(resultKey, ast.NewTerm(results))
+	}
+
+	if len(e.Erased) > 0 {
+		erased, err := ast.InterfaceToValue(e.Erased)
+		if err != nil {
+			return nil, err
+		}
+		event.Insert(erasedKey, ast.NewTerm(erased))
+	}
```

Comment:
Since we're including "erased" should we also include "masked"? It seems like this is going to get out-of-sync quite easily; a big comment on the event structure would be helpful. It would be good to mention what fields are excluded (if any currently).

Reply:
Ah yea, that slipped by as the changes for "masked" came in while these changes were a WIP. Fixed in the latest changes and I added a big warning to the struct definition.